### PR TITLE
Floating table

### PR DESCRIPTION
**Pull request by AaronGullickson**
# Summary

This PR addresses issue #157, specifically the ability to create latex floating environment tables with the use of the `table` + `tabular` environments rather than the `longtable` environment. Currently, I believe this is a major limitation of `gt` because tables are often broken up across pages, and most people expect tables to operate in a similar manner to figures. 

I added an option to `tab_option` called `latex.use.longtable` which defaults to FALSE. I then use a few `ifelse` clauses in `utils_render_latex.R` to get the proper output. Specification of the positioning of the float works well from the quarto code chunk option `tbl-pos`. 

If I run the following code:

```r
head(state.x77) |> 
  as.data.frame() |>
  gt() |>
  tab_options(latex.use.longtable = FALSE) |>
  as_latex()
```

Here is the latex output with the option set to FALSE:

```
"\\begin{table}\n\\begin{tabular}{rrrrrrrr}\n\\toprule\nPopulation & Income & Illiteracy & Life Exp & Murder & HS Grad & Frost & Area \\\\ \n\\midrule\\addlinespace[2.5pt]\n3615 & 3624 & 2.1 & 69.05 & 15.1 & 41.3 & 20 & 50708 \\\\ \n365 & 6315 & 1.5 & 69.31 & 11.3 & 66.7 & 152 & 566432 \\\\ \n2212 & 4530 & 1.8 & 70.55 & 7.8 & 58.1 & 15 & 113417 \\\\ \n2110 & 3378 & 1.9 & 70.66 & 10.1 & 39.9 & 65 & 51945 \\\\ \n21198 & 5114 & 1.1 & 71.71 & 10.3 & 62.6 & 20 & 156361 \\\\ \n2541 & 4884 & 0.7 & 72.06 & 6.8 & 63.9 & 166 & 103766 \\\\ \n\\bottomrule\n\\end{tabular}\n\\end{table}"
```

Here is the latex output with the option set to TRUE:

```
"\\begin{longtable}{rrrrrrrr}\n\\toprule\nPopulation & Income & Illiteracy & Life Exp & Murder & HS Grad & Frost & Area \\\\ \n\\midrule\\addlinespace[2.5pt]\n3615 & 3624 & 2.1 & 69.05 & 15.1 & 41.3 & 20 & 50708 \\\\ \n365 & 6315 & 1.5 & 69.31 & 11.3 & 66.7 & 152 & 566432 \\\\ \n2212 & 4530 & 1.8 & 70.55 & 7.8 & 58.1 & 15 & 113417 \\\\ \n2110 & 3378 & 1.9 & 70.66 & 10.1 & 39.9 & 65 & 51945 \\\\ \n21198 & 5114 & 1.1 & 71.71 & 10.3 & 62.6 & 20 & 156361 \\\\ \n2541 & 4884 & 0.7 & 72.06 & 6.8 & 63.9 & 166 & 103766 \\\\ \n\\bottomrule\n\\end{longtable}\n"
```

I have also supplied an example qmd document in `tests/gt-examples/03-latex/latex-15-floating.qmd`. 

I am submitting this PR currently as a draft as I have not fully stress tested it nor have I written a proper `testthat` test (not very experienced with those). I am happy to proceed with doing so, if the maintainers think I am on the right path here, but wanted to get feedback before continuing.

# Related GitHub Issues and PRs

- Ref: #157

# Checklist

- [x] I understand and agree to the [Code of Conduct](https://www.contributor-covenant.org/version/2/1/code_of_conduct.html).
- [x] I have listed any major changes in the [NEWS](https://github.com/rstudio/gt/blob/master/NEWS.md).
- [x] I have added [`testthat`](https://github.com/r-lib/testthat) unit tests to [`tests/testthat`](https://github.com/rstudio/gt/tree/master/tests/testthat) for any new functionality.

Fixes: https://github.com/rstudio/gt/issues/157

### CODE REVIEW

**Comment by AaronGullickson** [comment]
Correction, I did not commit the qmd file because these are apparently ignored by the project.

**Comment by AaronGullickson** [comment]
Commit 51da31ed02ea03e7a9cab77d3e56cc46937e6e60 add tests on the latex format for the new case.

**Comment by rich-iannone** [comment]
@AaronGullickson Apologies for not commenting earlier but thank you for the work done here! We have another PR underway that focuses on LaTeX output (https://github.com/rstudio/gt/pull/1595). It would be good to have that merged first before reviewing this one. (After that one is merged, the merge conflict here could be also addressed.)

**Comment by AaronGullickson** [comment]
@rich-iannone Sounds good. One important thing to consider when reviewing is which case (floating table or longtable) should be the default. I set it up here with floating table as the default, because I think that will be a more common case - i.e. if a table is not too long for a page, it should float by default. However, this will be a breaking change for existing documents in the sense that someone with a multipage table will now have it running off the page. I think its worth it for the long term benefit, but wanted to raise the issue here.

**Comment by AaronGullickson** [comment]
I am actually going to make one more change - I want to see if I can give the user the ability to specify the float position with a table option, rather than depend on Quarto.

**Comment by AaronGullickson** [comment]
Ok, never mind  about the floating position. When using captions from a code chunk, pandoc injects caption information into the table immediately after the curly closing bracket of \begin{table} which breaks the pre-existing floating position and you just end up with unintentional text in the table. So its better to specify floating position with the code chunk option `tbl-pos` which works correctly.

I added a news item and all checks are passing. It should be ready for review.

**Comment by kbrevoort** [comment]
@AaronGullickson May I ask a question about this PR as an interested user who would love for {gt} to include a float option?  

I believe that the `tbl-pos` chunk option only works in Quarto.  That won't be available to users who want to keep working with RMarkdown or people who want to use {gt} to generate tables that they can upload to Overleaf or use with a local install of TeX Live.  Is it possible for those users to specify a floating position without editing the code by hand each time they generate a table?

**Comment by AaronGullickson** [comment]
@kbrevoort Thats a good point. I did try to add that as another `tab_option` but found that it will not work correctly in Quarto. Let my try it again and see if it works in R Markdown. I can just add some documentation to the option to tell Quarto users not to do it this way, I suppose.

**Comment by rich-iannone** [comment]
Just wanted to chime in here and offer that the `check_quarto()` function could be used at render time to determine whether the rendering is occurring in a Quarto environment. This could be used in a conditional to enable options that are specific to Quarto.

**Comment by AaronGullickson** [comment]
I added a `latex.tbl.pos` option that can be used for floating tables. I used the `check_quarto()` function to ensure it does not mess up quarto output. I think the PR should be ready unless you think there is anything more it needs.

**Comment by AaronGullickson** [comment]
Fixed a couple of issues, but some tests are failing in an area I didn't touch so not sure what is going on there.

**Comment by AaronGullickson** [comment]
Also not sure where to do the license thing.

**Comment by AaronGullickson** [comment]
Fixed it. Forgot tests checked the total number of options. Had to increment expectations.

**Comment by kuriwaki** [comment]
Thank you @AaronGullickson.  This new option `latex.use.longtable = TRUE` removes the `\begin{table}` environment (which is great) and changes `tabular*` into a `longtable` tabular. Would it be straightforward to add a similar option, something like `latex.float` (that defaults to TRUE), which, if set to FALSE, will simply remove the `\begin{table}` and caption but keep the `tabular*`? 

In #157, I was thinking of use-cases where it's up to the user to generate the correct captions and environment for a non-floating table.

**Comment by AaronGullickson** [comment]
> Thank you @AaronGullickson. This new option `latex.use.longtable = TRUE` removes the `\begin{table}` environment (which is great) and changes `tabular*` into a `longtable` tabular. Would it be straightforward to add a similar option, something like `latex.float` (that defaults to TRUE), which, if set to FALSE, will simply remove the `\begin{table}` and caption but keep the `tabular*`?
> 
> In #157, I was thinking of use-cases where it's up to the user to generate the correct captions and environment for a non-floating table.

It should be easier to set something like this once this PR is merged.

**Comment by nielsbock** [comment]
The option to convert from a longtable to a table environment is a great feature that will make gt much more compatible with latex! I've noticed that adding a title using `tab_header()` will cause an error when knitting to pdf. Inserting a title:

```{r}
head(state.x77) |> 
  as.data.frame() |>
  gt() |>
  tab_options(latex.use.longtable = FALSE) |> 
  tab_header(title = "this is a title") |>
  gtsave("latex-table.tex")
```
essentially creates an unnumbered caption that is placed within the tabular environment which causes the error:

``` {=tex}
\begin{table}[!t]
\fontsize{12.0pt}{14.4pt}\selectfont
\begin{tabular*}{\linewidth}{@{\extracolsep{\fill}}rrrrrrrr}
\caption*{
{\large this is a title}
} \\ 
\toprule
Population & Income & Illiteracy & Life Exp & Murder & HS Grad & Frost & Area \\ 
\midrule\addlinespace[2.5pt]
3615 & 3624 & 2.1 & 69.05 & 15.1 & 41.3 & 20 & 50708 \\ 
365 & 6315 & 1.5 & 69.31 & 11.3 & 66.7 & 152 & 566432 \\ 
2212 & 4530 & 1.8 & 70.55 & 7.8 & 58.1 & 15 & 113417 \\ 
2110 & 3378 & 1.9 & 70.66 & 10.1 & 39.9 & 65 & 51945 \\ 
21198 & 5114 & 1.1 & 71.71 & 10.3 & 62.6 & 20 & 156361 \\ 
2541 & 4884 & 0.7 & 72.06 & 6.8 & 63.9 & 166 & 103766 \\ 
\bottomrule
\end{tabular*}
\end{table}
```
placing the caption just outside the tabular environment, for instance just after`\begin{table}` solves the issue:

``` {=tex}
\begin{table}[!t]
\caption*{
{\large this is a title}
}
\fontsize{12.0pt}{14.4pt}\selectfont
\begin{tabular*}{\linewidth}{@{\extracolsep{\fill}}rrrrrrrr}
\toprule
Population & Income & Illiteracy & Life Exp & Murder & HS Grad & Frost & Area \\ 
\midrule\addlinespace[2.5pt]
3615 & 3624 & 2.1 & 69.05 & 15.1 & 41.3 & 20 & 50708 \\ 
365 & 6315 & 1.5 & 69.31 & 11.3 & 66.7 & 152 & 566432 \\ 
2212 & 4530 & 1.8 & 70.55 & 7.8 & 58.1 & 15 & 113417 \\ 
2110 & 3378 & 1.9 & 70.66 & 10.1 & 39.9 & 65 & 51945 \\ 
21198 & 5114 & 1.1 & 71.71 & 10.3 & 62.6 & 20 & 156361 \\ 
2541 & 4884 & 0.7 & 72.06 & 6.8 & 63.9 & 166 & 103766 \\ 
\bottomrule
\end{tabular*}
\end{table}
```
If this is a known issue, please mark my comment as outdated.

**Comment by olivroy** [comment]
@AaronGullickson if you want, could you fix merge conflicts. If @rich-iannone agrees, we could take this early for the upcoming 0.12.0 release to be able to test it more thoroughly

File: R/utils_render_latex.R
Comment (olivroy) [inline R/utils_render_latex.R:1337]:
```suggestion
  if (table_width == 'auto' ||
```
condition

File: tests/gt-examples/03-latex/latex-13-adding_footnotes.Rmd
Comment (olivroy) [inline tests/gt-examples/03-latex/latex-13-adding_footnotes.Rmd:38]: Why did you comment that out?

File: tests/testthat/test-as_latex.R
Comment (olivroy) [inline tests/testthat/test-as_latex.R:47]:
```suggestion
    tab_options(table.width = "600px", latex.use.longtable = FALSE) %>%
```

Nitpick: We use double quotes in the codebase!

File: tests/testthat/test-as_latex.R
Comment (olivroy) [inline tests/testthat/test-as_latex.R:174]: Idem here! You can use styler to restyle the selection in RStudio. https://styler.r-lib.org/

**Review by olivroy — COMMENTED** [review]
Thanks for working on this! A few comments.


Also, I tested this locally with `usethis::pr_fetch(1588)`.

There is something I don't understand.

Given the following code (Rmd):


````Rmd
---
title: "latex-03-pressure"
output: 
  - pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
devtools::load_all(".")
```

Create a display table based on `pressure`.

```{r}
pressure %>% dplyr::bind_rows(pressure) %>% dplyr::bind_rows(pressure) %>% 
gt() %>%
  fmt_scientific(
    columns = pressure,
    decimals = 2
  ) %>% 
  tab_options(latex.use.longtable = TRUE)
```

I notice that the table headings aren't repeated as the table crosses 2 pages. 
![image](https://github.com/rstudio/gt/assets/52606734/7d294501-a5a3-4ce0-8bc5-5b5f9c306396)
````

I haven't followed the full conversation, so please enlighten me. In the [longtable manual](https://mirrors.ibiblio.org/CTAN/macros/latex/required/tools/longtable.pdf), this seems like a supported feature.

**Comment by olivroy** [comment]
Maybe some \endhead \endfirsthead \endfoot \endlastfoot could be placed strategically From Ch 3 https://mirrors.ibiblio.org/CTAN/macros/latex/required/tools/longtable.pdf 

I don't know LaTeX that much, and this feels a bit like trial and error to me.

**Comment by kbrevoort** [comment]
@olivroy There are a couple of open issues about headers & captions not repeating on tables that span multiple pages (#1630, #1061).  The reason I haven't submitted a PR on that myself is my sense is that, while having the column headers repeat on each page would be very easy, repeating the captions (i.e., "Table 1:  My Table Title (continued)" on subsequent pages) would be trickier.  I'm confident it's doable (and if someone wants to get to it before I do, great!) but it's probably worthy of its own PR.

It's also a very different issue than what @AaronGullickson is addressing.  This PR provides an alternative environment to longtable with many benefits for users (myself included).  But one thing the new floating environment can't do is span multiple pages.  It's not my call, but I recommend not holding up this PR because it doesn't fix an unrelated issue.  (We can work on #1630 and #1061, and hopefully, the change that @kuriwaki was asking for while people kick the tires of the new environment.)

I hope that helps. @AaronGullickson, thanks for implementing a floating environment. I'm looking forward to trying it out once it's merged!

File: tests/gt-examples/03-latex/latex-13-adding_footnotes.Rmd
Comment (rich-iannone) [reply tests/gt-examples/03-latex/latex-13-adding_footnotes.Rmd:38]: I looked into this and if the line is uncommented *and* `tab_options(latex.use.longtable = TRUE)` is removed, we get a LaTeX error when rendering the .Rmd file:

```
! You can't use `\hrule' here except with leaders.
\caption@hrule ->\hrule 
                        \@height \z@ 
l.199 }
        \\ 

Error: LaTeX failed to compile latex-13-adding_footnotes.tex. See https://yihui.org/tinytex/r/#debugging for debugging tips. See latex-13-adding_footnotes.log for more info.
Execution halted
```

**Comment by olivroy** [comment]
@kbrevoort thanks for the explanations!

@AaronGullickson thanks for sticking with it

@rich-iannone I recommend merging this and possibly have longtable be the default at some point. The fact that long tables can now be broken up across pages instead of overflowing is a nice improvement!

File: tests/gt-examples/03-latex/latex-13-adding_footnotes.Rmd
Comment (olivroy) [reply tests/gt-examples/03-latex/latex-13-adding_footnotes.Rmd:38]: @AaronGullickson Are you able to fix this? It is important that LaTeX tables don't get compilation issues if they have a header

Comment (nielsbock) [reply tests/gt-examples/03-latex/latex-13-adding_footnotes.Rmd:38]: The error is due to the unnumbered caption being placed within the tabular environment when `tab_options(latex.use.longtable = FALSE)`. I'm not exactly sure how why the error occurs in the test because this specific problem is solved with the code in 727-762 in `R/export.R`.

**Comment by rich-iannone** [comment]
@AaronGullickson if you could make that `tab_header()` fix, then I think we could proceed with merging this PR.

**Comment by AaronGullickson** [comment]
Thanks for all of this helpful feedback. I am away for a few days built will make the corrections first thing Monday.

File: tests/testthat/test-tab_footnote.R
Comment (olivroy) [inline tests/testthat/test-tab_footnote.R:784]: In test-tab_footnote and test-tab_stub_indent, could you remove the changes? We are testing html output here. If you want to add more tests for latex output, please do so in test-l_* .

Generally, please remove the changes where you are not testing for LaTeX output specifically, and feel free to create new tests for things you may want to test

**Comment by olivroy** [comment]
@AaronGullickson when you take this on, please merge my PR first https://github.com/AaronGullickson/gt/pull/4. I took care of resolving some merge conflicts.

File: tests/testthat/test-tab_footnote.R
Comment (rich-iannone) [reply tests/testthat/test-tab_footnote.R:784]: Regarding the option `latex.use.longtable`, could we rename this slightly to `latex.use_longtable`? This is just for consistency with the idea that dots separate either locations, contexts, or actions. It seems to me that `use_longtable` is a single concrete action here.

File: tests/gt-examples/03-latex/latex-13-adding_footnotes.Rmd
Comment (AaronGullickson) [reply tests/gt-examples/03-latex/latex-13-adding_footnotes.Rmd:38]: Yes, I must have commented that out early on and forgot about it. However, when I uncomment the line, the Rmd file renders fine for me with the latest changes from the PR. @rich-iannone Can you please confirm it gives you that error with the latest?

**Comment by AaronGullickson** [comment]
I think I have made all the requested changes. I changed the option name from `latex.use.longtable` to `latex.use_longtable` . I also uncommented the line I commented out in one of the Rmd files. It renders fine for me locally using either the `table` or `longtable` environment. 

One last issue to consider is the default environment. Currently, this PR uses a floating table environment as the default. I would argue that this is the best approach because tables that are not longer than a page should be floated for the best results. Thus, longtable is only really needed for long tables, as the name implies. This does however change the environment from its previous default (and only environment) which was longtable, which could cause some confusion when users re-render existing documents (for example).

**Review by rich-iannone — APPROVED** [review]

**Comment by rich-iannone** [comment]
Thanks again @AaronGullickson for all your time and patience with this. We're finally going to merge this PR; will do so after all CI runs pass!

File: tests/gt-examples/03-latex/latex-13-adding_footnotes.Rmd
Comment (rich-iannone) [reply tests/gt-examples/03-latex/latex-13-adding_footnotes.Rmd]: @AaronGullickson everything is good here now!

**Pull request merged**
